### PR TITLE
Feat: Add a multi-select on relation filters

### LINKED ISSUE #1719: Add a multi-select on relation filters
## Scope & Context

In every filter dropdown we have in the app.

## Current behavior

We can only select one relation in a filter : 

image

## Expected behavior

We would like to be able to select multiple filters, for example I'd like to create a view "FAANG" in my people table with : Google, Microsoft, Facebook, etc.

## Technical inputs

We can use the same picker we have in the right drawer for relations : 

image

We need to modify a little bit our filters structure : 

We can rename `value` to `orValues` or something like that so we know it will always be applied as an OR clause. (AND in the same filter seems less useful and could be another feature if needed)

```tsx
export type Filter = {
  key: string;
  type: FilterType;
  orValues: string[]; // Use an array of string by default to create an OR clause ?
  displayValue: string;
  displayAvatarUrl?: string;
  operand: FilterOperand;
};
```

### PR DESCRIPTION
**Pull request by AyushAgrawal-A2**
Closes #1719

- Create multi-select filters with operand select for the user, people & company entity.

https://github.com/twentyhq/twenty/assets/54364088/57f34179-3027-4b69-882b-6aaecf6ac800

### CODE REVIEW

**Review by lucasbordeau — CHANGES_REQUESTED** [review]

File: front/src/modules/ui/view-bar/components/FilterDropdownMultipleEntitySearchSelect.tsx
Comment (AyushAgrawal-A2) [reply front/src/modules/ui/view-bar/components/FilterDropdownMultipleEntitySearchSelect.tsx:30]: done

**Comment by lucasbordeau** [comment]
I'm not able to see the changes you made, could you merge main and push ?

**Comment by charlesBochet** [comment]
@Bonapara does design looks good to you? :)

**Comment by AyushAgrawal-A2** [comment]
> I'm not able to see the changes you made, could you merge main and push ?

Found a bug while screen recording. Just fixed it.

File: front/src/modules/ui/view-bar/components/FilterDropdownMultipleEntitySearchSelect.tsx
Comment (lucasbordeau) [inline front/src/modules/ui/view-bar/components/FilterDropdownMultipleEntitySearchSelect.tsx:52]: We avoid generic names even in lambdas, better to have `previousSelectedEntities`

File: front/src/modules/ui/view-bar/components/FilterDropdownMultipleEntitySearchSelect.tsx
Comment (lucasbordeau) [inline front/src/modules/ui/view-bar/components/FilterDropdownMultipleEntitySearchSelect.tsx:54]: `enitity`

**Comment by lucasbordeau** [comment]
@AyushAgrawal-A2 

- Can you see if `OverflowingTextWithTooltip` can fit your needs ? Using this component + a max-width on the chip ?
- The API is way cleaner, nice !
- Is it difficult for you to change the filter at every select/unselect, not after we exit the dropdown ? We can always do it in another ticket.
- We should erase the selected values when closing the dropdown. => Meaning we should reset the state of the multiple selection so we don't have the previous selection shown in the dropdown, but it seems to be fixed with your new changes !

**Comment by AyushAgrawal-A2** [comment]
> * Is it difficult for you to change the filter at every select/unselect, not after we exit the dropdown ? We can always do it in another ticket.

I actually first implemented per-click update. On every selection, the table was getting refreshed and the menu item order was also getting changed as the selected item would move to the top. I didn't like the user experience while testing it.
Can revert back if that is the desired behavior.

**Comment by Bonapara** [comment]
> @Bonapara
> 
> 1. I have set the display name for multi-select filter as a comma-separated list. This will get too long.
>    image
>    Options:
>    
>    * Trim it after a max length "Facebook, Google, Air..."
>    * Show full name if only one is selected and number if many e.g. "3 Selected"
>    * ??
> 2. Operand name in Operand Select menu
>    image
> 3. Operand in Filter chip. "Company: NotIn ..." i.e. `: In` `: NotIn` part. I feel `: NotIn` is not readable.

Hi @AyushAgrawal-A2 

- For your first question, let's do you second suggestion → Show full name if only one is selected and number if many e.g. "3 Selected"
- For your second question I think we can keep the behavior of people operand name : "is" and "not" (without displaying "is" in the chip)

**Comment by lucasbordeau** [comment]
> > * Is it difficult for you to change the filter at every select/unselect, not after we exit the dropdown ? We can always do it in another ticket.
> 
> I actually first implemented per-click update. On every selection, the table was getting refreshed and the menu item order was also getting changed as the selected item would move to the top. I didn't like the user experience while testing it. Can revert back if that is the desired behavior.

Good point, what are your thoughts on that @Bonapara, I feel like it's better to show changes on each click.

**Comment by AyushAgrawal-A2** [comment]
@lucasbordeau @Bonapara 
- Made all changes in "as per feedback" commit [fcb51a6](https://github.com/twentyhq/twenty/pull/1822/commits/fcb51a6e122cf57ba0473ce16b89ab96dde95fc3)
- Switched to per click update in last commit [364809e](https://github.com/twentyhq/twenty/pull/1822/commits/364809e114fa15e2629294e5be2daa522acb28cc)

Please try it and decide.

P.S.: storybook fix is pending, and can be fixed only after behavior is finalized

**Comment by Bonapara** [comment]
> > > * Is it difficult for you to change the filter at every select/unselect, not after we exit the dropdown ? We can always do it in another ticket.
> > 
> > 
> > I actually first implemented per-click update. On every selection, the table was getting refreshed and the menu item order was also getting changed as the selected item would move to the top. I didn't like the user experience while testing it. Can revert back if that is the desired behavior.
> 
> Good point, what are your thoughts on that @Bonapara, I feel like it's better to show changes on each click.

I agree with your understanding. Implementing changes on each click could enhance the experience, but we should prevent the menu from reordering after each selection. The reordering should only occur when the dropdown is closed and then reopened. We should open a new issue for that.

**Comment by lucasbordeau** [comment]
I'll let @charlesBochet review because I'm quite busy but seems ok for me !

**Comment by AyushAgrawal-A2** [comment]
@Bonapara 

> but we should prevent the menu from reordering after each selection. The reordering should only occur when the dropdown is closed and then reopened. We should open a new issue for that.

This will require refactoring and reviewing the same code again, Please let me fix this in this PR itself.

**Comment by AyushAgrawal-A2** [comment]
@lucasbordeau @charlesBochet @Bonapara 
Changes:-
- Fixed menu item reorder on item select/deselect.
- Fixed correct operand select on filter menu open.


https://github.com/twentyhq/twenty/assets/54364088/e53310ef-5240-4985-aa64-5d723b8cc483

Submitting again for review.

**Comment by lucasbordeau** [comment]
Waiting for relation field type refactor with new flexible backend.

**Comment by lucasbordeau** [comment]
Hi @AyushAgrawal-A2, thank you very much for your efforts. 

After waiting for a long time, we have decided to implement this feature with the core team because it touches a lot of moving parts of the new architecture, in https://github.com/twentyhq/twenty/pull/2858.

Please feel free to tell us anything we might have missed in our implementation.

**Pull request closed without merging**
